src/Analyzers/Core/Analyzers/RemoveRedundantEquality/RemoveRedundantEqualityDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis.RemoveRedundantEquality
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal sealed class RemoveRedundantEqualityDiagnosticAnalyzer
+        : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public RemoveRedundantEqualityDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.RemoveRedundantEqualityDiagnosticId,
+                   option: null,
+                   new LocalizableResourceString(nameof(AnalyzersResources.Remove_redundant_equality), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterOperationAction(AnalyzeBinaryOperator, OperationKind.BinaryOperator);
+
+        private void AnalyzeBinaryOperator(OperationAnalysisContext context)
+        {
+            var operation = (IBinaryOperation)context.Operation;
+            if (operation.OperatorMethod is not null)
+            {
+                // We shouldn't report diagnostic on overloaded operator as the behavior can change.
+                return;
+            }
+
+            if (operation.OperatorKind is not (BinaryOperatorKind.Equals or BinaryOperatorKind.NotEquals))
+            {
+                return;
+            }
+
+            var rightOperand = operation.RightOperand;
+            var leftOperand = operation.LeftOperand;
+            // The reported additionalLocations represents the non-redundant side of equality.
+            // For example: if (x == true) {}
+            //     additionalLocations: x
+            //     primary Location:    x == true
+            // additionalLocations is used by the codefix as the location of the replacement node.
+            // So, the code fix will do: Replace(node at diagnostic.Location, node at diagnostic.AdditionalLocations[0]).
+            if (rightOperand.ConstantValue.HasValue && rightOperand.ConstantValue.Value is true)
+            {
+                context.ReportDiagnostic(DiagnosticHelper.Create(Descriptor,
+                    rightOperand.Syntax.GetLocation(),
+                    ReportDiagnostic.Hidden,
+                    additionalLocations: new[] { leftOperand.Syntax.GetLocation() },
+                    properties: null));
+            }
+            else if (leftOperand.ConstantValue.HasValue && leftOperand.ConstantValue.Value is true)
```

Comment:
note: this will mean that if someone has `const bool Activated = true; ... if (state == Activated)`  then we will suggest simplifying things.  I don't think we want that.  this should really be about a direct test against hte `true/false` literals.

Reply:
@CyrusNajmabadi Do you think I should completely give up using RegisterOperationAction and just do the analysis based on the syntax tree instead?

I'm not sure, but I think it'll make things much easier.

Reply:
up to you.  syntax might be easier here, or it might not be.  i wouldn't be sure without trying both myself

Reply:
You don't need to switch to syntax for such a test. You can add an extra condition `leftOperation.Kind == OperationKind.Literal`

Reply:
@mavasani Thanks. Is there a way to get the operand location without switching to syntax?

Reply:
You can possibly get the syntax node for left operand, fetch its last token and then invoke GetNextToken?

Reply:
i would not do this.  for one thing, we explicitly state that there is not necessairly a relation between IOp and syntax.  So this analyzer should both check for implicit IOps and also vlaidate that it's on hte syntactic form it expects.

Reply:
I came up with `operation.Syntax.SyntaxTree.FindTokenOnLeftOfPosition(rightOperand.Syntax.SpanStart, context.CancellationToken).GetLocation(),` for the operator location, but not sure if you'll like it.

Reply:
i do not like it.  just actually test the node types, and use the syntax. if you want to share common code, use ISyntaxFacts.

Reply:
You can use existing API `ISyntaxFacts.GetPartsOfBinaryExpression`